DM-30895: Move AssembleCoaddTask and its children from pipe_tasks

arunkannawadi opened the pull request.

fred3m commented on python/lsst/drp/tasks/assemble_coadd.py:
```diff
@@ -0,0 +1,1777 @@
+    @staticmethod
+    def _subBBoxIter(bbox, subregionSize):
+        """Iterate over subregions of a bbox.
+
+        Parameters
+        ----------
+        bbox : `lsst.geom.Box2I`
+            Bounding box over which to iterate.
+        subregionSize : `lsst.geom.Extent2I`
+            Size of sub-bboxes.
+
+        Yields
+        ------
+        subBBox : `lsst.geom.Box2I`
+            Next sub-bounding box of size ``subregionSize`` or smaller; each
+            ``subBBox`` is contained within ``bbox``, so it may be smaller than
+            ``subregionSize`` at the edges of ``bbox``, but it will never be
+            empty.
+
+        Raises
+        ------
+        RuntimeError
+            Raised if any of the following occur:
+            - The given bbox is empty.
+            - The subregionSize is 0.
+        """
+        if bbox.isEmpty():
+            raise RuntimeError("bbox %s is empty" % (bbox,))
+        if subregionSize[0] < 1 or subregionSize[1] < 1:
+            raise RuntimeError("subregionSize %s must be nonzero" % (subregionSize,))
+
+        for rowShift in range(0, bbox.getHeight(), subregionSize[1]):
+            for colShift in range(0, bbox.getWidth(), subregionSize[0]):
+                subBBox = geom.Box2I(bbox.getMin() + geom.Extent2I(colShift, rowShift), subregionSize)
+                subBBox.clip(bbox)
+                if subBBox.isEmpty():
+                    raise RuntimeError("Bug: empty bbox! bbox=%s, subregionSize=%s, "
+                                       "colShift=%s, rowShift=%s" %
+                                       (bbox, subregionSize, colShift, rowShift))
+                yield subBBox
```
This is in `lsst.pipe.tasks.coaddBase`, so it can be removed as a static method of this class.

fred3m reviewed (changes requested):
Just a few small comments and corrections.

arunkannawadi commented:
Great catches, Fred! I've addressed the comments now.

 - [ ] Move assembleChi2Coadd.py
 - [ ] Remove the subBBoxIter static method (and use the one from coaddBase)
 - [ ]  Add a period of kernel in a docstring
